safe_control_gym/safety_filters/cbf/cbf.py

```diff
@@ -0,0 +1,327 @@
+            uncertified_action (ndarray): The uncertified_controller's action.
+            info (dict): The info at this timestep.
+
+        Returns:
+            certified_action (ndarray): The certified action.
+            success (bool): Whether the safety filtering was successful or not.
+        '''
+
+        self.results_dict['uncertified_action'].append(uncertified_action)
+        certified_action, success = self.solve_optimization(current_state, uncertified_action)
+        self.results_dict['feasible'].append(success)
+        self.results_dict['certified_action'].append(certified_action)
+        self.results_dict['correction'].append(np.linalg.norm(certified_action-uncertified_action))
+
+        return certified_action, success
+
+    def is_cbf(self, num_points=100, tolerance=0.01):
+        '''
+        Check if the provided CBF candidate is actually a CBF for the system using a gridded approach.
+
+        Args:
+            num_points (int): The number of points in each dimension to check.
+            tolerance (float): The tolerance of the condition outside the superlevel set.
+
+        Returns:
+            valid_cbf (bool): Whether the provided CBF candidate is valid.
+            infeasible_states (list): List of all states for which the QP is infeasible.
+        '''
+        valid_cbf = False
+
+        # Select the states to check the CBF condition
+        max_bounds = np.array(self.state_limits)
+        # Add some tolerance to the bounds to also check the condition outside of the superlevel set
+        max_bounds += tolerance
+        min_bounds = -max_bounds
+
+        # state dimension and input dimension
+        nx, nu = self.model.nx, self.model.nu
+
+        # Make sure that every vertex is checked
+        num_points = max(2 * nx, num_points + num_points % (2 * nx))
+        num_points_per_dim = num_points // nx
+
+        # Create the lists of states to check
+        states_to_sample = [np.linspace(min_bounds[i], max_bounds[i], num_points_per_dim) for i in range(nx)]
+        states_to_check = cartesian_product(*states_to_sample)
+
+        # Set dummy control input
+        control_input = np.ones((nu, 1))
+
+        num_infeasible = 0
+        num_infeasible_states_inside_set = 0
+        infeasible_states = []
+
+        # Check if the optimization problem is feasible for every considered state
+        for state in states_to_check:
+            # Certify action
+            _, success = self.certify_action(state, control_input)
+
+            if not success:
+                infeasible_states.append(state)
+                num_infeasible += 1
+                barrier_at_x = self.cbf(X=state)['cbf']
+
+                # Check if the infeasible point is inside or outside the superlevel set. Note that the sampled region makes up a
+                # box, but the superlevel set is not. The superlevel set only needs to be contained inside the box.
+                if barrier_at_x > 0:
```

Comment:
Maybe something like `barrier_at_x >= 0.0 + eps` where eps is some positive number is more explicit?